### PR TITLE
Fix gradient of tf.nn.isotonic_regression for 1-D inputs

### LINKED ISSUE #127234: tf.nn.isotonic_regression returns incorrect gradient for a pooled block
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13.5

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.isotonic_regression` returns an incorrect first derivative for a stable pooled block.

The reproducer evaluates decreasing isotonic regression on `[4, t, 2]` with `t = 0.7`. For `0 < t < 2`, the last two entries form a stable pooled block with value `(t + 2) / 2`. Therefore, the selected output `y[1]` has derivative `1/2` throughout an open neighborhood.

The forward value is computed correctly as `1.35`, but reverse-mode autodiff returns `1.0` for the derivative with respect to `t`. The expected derivative is `0.5`.

Expected behavior?

The first derivative of the selected `tf.nn.isotonic_regression` output at `t = 0.7` should be close to `0.5`, not `1.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    x = tf.stack([
        tf.constant(4, dtype=t.dtype),
        t,
        tf.constant(2, dtype=t.dtype),
    ])
    y, segments = tf.nn.isotonic_regression(
        inputs=x,
        decreasing=True,
        axis=0,
    )
    return y[1]

x = tf.constant(0.7, dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    y = target(x)

actual = tape.gradient(y, x)
expected = 0.5

print("forward:", y.numpy())
print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.nn.isotonic_regression returns incorrect gradient for a pooled block")
else:
    print("not reproduced")
```

### Relevant log output

```shell
forward: 1.35
actual: 1.0
expected: 0.5
BUG REPRODUCED: tf.nn.isotonic_regression returns incorrect gradient for a pooled block
```

### PR DESCRIPTION
**Pull request by Phoebus-Liu**
`tf.nn.isotonic_regression` returns an incorrect gradient for 1-D inputs.
Reported in #127234.

Pooling happens along the last axis, and `_MeanAggregator` in `nn_grad.py`
aggregates one row at a time by splitting its input along axis 0. For a 2-D
input axis 0 is the batch axis, so this is correct. For a 1-D input axis 0 is
the data axis instead, so every element became its own row: each segment
collapsed to a single element and the gradients of the values that were
pooled into a block were never aggregated.

For the case in the issue (`[4, t, 2]`, `decreasing=True`, `t = 0.7`, where
the last two elements are pooled into a block of value `(t + 2) / 2`) the
gradient of `y[1]` with respect to `t` is now `0.5` as expected, instead of
`1.0`.

The gradient shape is fixed as well. For a `(50,)` input the gradient used to
be `(50, 1)`, which makes `gradient_checker_v2.compute_gradient` fail
outright with `Incompatible shapes`; it is now `(50,)`.

A 1-D input is now expanded to a single row before the split and squeezed
back afterwards. The 2-D and higher-rank paths are unchanged.

Adds `testGradient1D`, which runs `gradient_checker_v2.compute_gradient` on a
1-D input (maximum deviation between analytical and numerical gradients:
9.464e-12). The existing `testGradientV2` only exercises a `(30, 50)` input,
so the 1-D path was never covered. The 2-D test still passes unchanged.

Fixes #127234

---

Note on the PyLint check: it reports two pre-existing `E1130` findings at
`tensorflow/python/ops/nn_test.py:879` and `:936`
(`return -np.sum(targets * np.log(pred + 1.0e-20), axis=1)`), both in the
existing `testSampledSoftmaxLoss` / `testSampledSoftmaxLossBf16` reference
implementations. Neither line is touched here - this PR only appends 18 lines
to `IsotonicTest` (~line 2054) and changes nothing else in that file.

The reports come from the toolchain rather than the code: the presubmit pins
`pylint==2.13.9` but leaves `numpy` unpinned, and on Python 3.9 that resolves
to numpy 2.0.2, where astroid 2.11.7 infers `np.sum(...)` as `NoneType`.
Reproduced locally with pylint 2.13.9 and the repo's own rcfile: numpy 2.0.2
-> 9.93/10 with exactly these two reports, numpy 2.5.3 -> 10.00/10. Python
version is not a factor, the same split reproduces on 3.12.

Filed as #127559 and fixed separately in #127561, rather than expanding the
scope of this bugfix.

### CODE REVIEW

**Comment by Phoebus-Liu** [comment]
The PyLint failure here is pre-existing and unrelated to this change.

The two `E1130` reports point at `tensorflow/python/ops/nn_test.py:879` and `:936`, both `return -np.sum(targets * np.log(pred + 1.0e-20), axis=1)` inside the existing `testSampledSoftmaxLoss` / `testSampledSoftmaxLossBf16` helpers. This PR only adds 18 lines at the end of `IsotonicTest` (~line 2054) and deletes nothing, so neither line is touched here.

The reports come from the toolchain, not from the code: the workflow pins `pylint==2.13.9` but leaves `numpy` unpinned, and on Python 3.9 that resolves to numpy 2.0.2, whose stubs make astroid 2.11.7 infer `np.sum(...)` as `NoneType`. Reproduced locally with pylint 2.13.9 and the repo's own rcfile:

- numpy 2.0.2 -> 9.93/10, exactly these two `E1130` reports
- numpy 2.5.3 -> 10.00/10, no reports

Python version is not a factor: the same split reproduces on 3.12.

I filed this as #127559 and will send the one-line fix separately, rather than expanding the scope of this bugfix.

**Review by dmiltr3 — APPROVED** [review]
LGTM! All internal TAP presubmit checks have passed successfully across Google infrastructure. Approving this PR.

**Comment by Phoebus-Liu** [comment]
Thank you, @dmiltr3, for the review and approval. I'm glad the internal TAP presubmit checks passed across the Google infrastructure. Thanks again for your time on this.
